File: packages/swift-sdk/SwiftExampleApp/SwiftExampleApp/Core/Services/TestnetFaucetService.swift
import CryptoKit
import Foundation

/// **Dev/QA tool only.** App-level client for the testnet DASH faucet at
/// `faucet.thepasta.org`. This is a plain `URLSession` HTTP client plus a
/// pure-Swift `cap.js` proof-of-work solver — it deliberately does **not**
/// touch the wallet / Platform FFI, the `platform-wallet` crate, or any
/// SDK business logic. It exists purely so QA can top up a fresh/drained
/// testnet wallet with one tap instead of round-tripping through the web
/// faucet (issue #3897).
///
/// The faucet enforces a server-side captcha (`cap.js`). The "soft"
/// challenge is a small proof-of-work (≈6.5M SHA-256 ops) that we solve
/// on-device off the main actor; the "hard" challenge returned on rate
/// limit is far too expensive to brute-force here, so on rate limit (or
/// any other failure) the caller falls back to opening the web faucet.

// MARK: - cap.js proof-of-work solver

enum CapSolver {
    /// cap.js seeded PRNG: FNV-1a-style seed + xorshift32. Emits `length`
    /// lowercase hex chars.
    ///
    /// Verified bit-identical to `@cap.js/widget` against the live server
    /// (cross-checked with `/tmp/cap_wrap_check.py` — both the wrapping and
    /// the masked int32 forms agree, and the wrapping form redeems live).
    static func prng(_ seed: String, _ length: Int) -> String {
        var state: UInt32 = 2166136261                 // 0x811C9DC5
        for u in seed.utf16 {                          // seeds are ASCII (hex token + digits [+ "d"])
            state ^= UInt32(u)
            let sh = (state << 1) &+ (state << 4) &+ (state << 7) &+ (state << 8) &+ (state << 24)
            state = state &+ sh
        }
        var out = ""
        while out.count < length {
            state ^= state << 13
            state ^= state >> 17
            state ^= state << 5
            out += String(format: "%08x", state)        // unsigned 32-bit, 8 hex chars
        }
        return String(out.prefix(length))
    }

    /// A single sub-challenge: brute-force the smallest non-negative
    /// integer `nonce` such that lowercase-hex(`SHA256(salt + String(nonce))`)
    /// starts with `target`.
    ///
    /// We never materialize the 64-char hex digest per attempt. `target`
    /// is parsed once into a byte prefix (plus an optional trailing high
    /// nibble when `target` has odd length) and compared directly against
    /// the raw `SHA256` digest bytes.
    static func solve(salt: String, target: String) -> Int {
        let targetChars = Array(target.utf8)
        let fullBytes = targetChars.count / 2
        let hasNibble = targetChars.count % 2 == 1

        // Parse the full leading bytes of the hex target.
        var prefix = [UInt8]()
        prefix.reserveCapacity(fullBytes)
        for i in 0..<fullBytes {
            let hi = hexNibble(targetChars[i * 2])
            let lo = hexNibble(targetChars[i * 2 + 1])
            prefix.append((hi << 4) | lo)
        }
        // Optional trailing high-nibble (odd-length target).
        let nibble: UInt8 = hasNibble ? hexNibble(targetChars[fullBytes * 2]) : 0

        let saltBytes = Array(salt.utf8)
        var nonce = 0
        while true {
            var hasher = SHA256()
            hasher.update(data: saltBytes)
            hasher.update(data: Array(String(nonce).utf8))
            let digest = hasher.finalize()

            if digestHasPrefix(digest, prefix: prefix, nibble: nibble, hasNibble: hasNibble) {
                return nonce
            }
            nonce += 1
        }
    }

    private static func digestHasPrefix(
        _ digest: SHA256.Digest,
        prefix: [UInt8],
        nibble: UInt8,
        hasNibble: Bool
    ) -> Bool {
        var idx = 0
        for byte in digest {
            if idx < prefix.count {
                if byte != prefix[idx] { return false }
                idx += 1
            } else if hasNibble {
                // Compare only the high nibble of the next digest byte.
                return (byte >> 4) == nibble
            } else {
                return true
            }
            if idx == prefix.count && !hasNibble {
                return true
            }
        }
        // Digest exhausted: matched iff we consumed the whole prefix and
        // there's no dangling nibble to check.
        return idx == prefix.count && !hasNibble
    }

    private static func hexNibble(_ c: UInt8) -> UInt8 {
        switch c {
        case 0x30...0x39: return c - 0x30            // '0'-'9'
        case 0x61...0x66: return c - 0x61 + 10       // 'a'-'f'
        case 0x41...0x46: return c - 0x41 + 10       // 'A'-'F'
        default: return 0
        }
    }
}

// MARK: - Outcome

/// Result of a faucet funding attempt. The view turns `.failed` /
/// `.rateLimited` into the clipboard + web-faucet fallback.
enum TestnetFaucetOutcome: Sendable {
    /// Funds were sent. `txid` is the Core transaction id; `amount` is the
    /// tDASH amount reported by `/api/status` (`coreFaucetAmount`).
    case sent(txid: String, amount: Double)
    /// Server rate-limited the request (429). Carries a user-facing message.
    case rateLimited(message: String)
    /// Any other failure (network error, solve failure, non-200 response).
    case failed(reason: String)
}

// MARK: - HTTP client

/// Thin async client for the testnet faucet. Stateless; create one per use.
struct TestnetFaucet {
    /// Faucet web host. Used both for the JSON API and as the web fallback URL.
    static let webURL = URL(string: "https://faucet.thepasta.org/")!

    private let host = URL(string: "https://faucet.thepasta.org")!
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Request 1 tDASH to `address`. Solves the soft captcha on-device and
    /// posts to `/api/core-faucet`. Safe to call from the main actor — the
    /// proof-of-work runs on a detached background-priority task.
    func requestCoreDash(address: String) async -> TestnetFaucetOutcome {
        let trimmed = address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return .failed(reason: "No address available")
        }

        // 1. Read status for the (rotatable) cap endpoint + faucet amount.
        let status: FaucetStatus
        do {
            status = try await fetchStatus()
        } catch {
            return .failed(reason: "Faucet status unavailable: \(error.localizedDescription)")
        }

        guard let capBase = URL(string: status.capEndpoint) else {
            return .failed(reason: "Invalid captcha endpoint")
        }

        // 2. Solve the soft captcha proof-of-work.
        let capToken: String
        do {
            capToken = try await solveCaptcha(capBase: capBase)
        } catch {
            return .failed(reason: "Captcha solve failed: \(error.localizedDescription)")
        }

        // 3. Post the funding request.
        return await postFaucet(
            address: trimmed,
            capToken: capToken,
            amount: status.coreFaucetAmount
        )
    }

    // MARK: status

    private struct FaucetStatus: Decodable {
        let capEndpoint: String
        let coreFaucetAmount: Double
    }

    private func fetchStatus() async throws -> FaucetStatus {
        let url = host.appendingPathComponent("api/status")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 30
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(FaucetStatus.self, from: data)
    }

    // MARK: captcha

    private struct ChallengeResponse: Decodable {
        struct Challenge: Decodable {
            let c: Int   // number of sub-challenges
            let s: Int   // salt hex length
            let d: Int   // target hex length
        }
        let challenge: Challenge
        let token: String
    }

    private struct RedeemResponse: Decodable {
        let success: Bool
        let token: String?
    }

    /// Runs the full cap.js handshake: request a challenge, brute-force the
    /// `c` sub-challenges (parallelized across cores), redeem for a capToken.
    private func solveCaptcha(capBase: URL) async throws -> String {
        // challenge
        let challengeURL = capBase.appendingPathComponent("challenge")
        let challenge: ChallengeResponse = try await postJSON(
            url: challengeURL,
            body: [:] as [String: String]
        )

        let c = challenge.challenge.c
        let s = challenge.challenge.s
        let d = challenge.challenge.d
        let token = challenge.token

        guard c > 0 else {
            throw FaucetError.message("Empty captcha challenge")
        }

        // Solve all c sub-challenges off the main actor, parallelized.
        let solutions: [Int] = try await withThrowingTaskGroup(
            of: (Int, Int).self
        ) { group in
            for i in 1...c {
                group.addTask(priority: .background) {
                    let salt = CapSolver.prng("\(token)\(i)", s)
                    let target = CapSolver.prng("\(token)\(i)d", d)
                    let nonce = CapSolver.solve(salt: salt, target: target)
                    return (i, nonce)
                }
            }
            var collected = Array(repeating: 0, count: c)
            for try await (i, nonce) in group {
                collected[i - 1] = nonce   // preserve sub-challenge order
            }
            return collected
        }

        // redeem
        let redeemURL = capBase.appendingPathComponent("redeem")
        let redeem: RedeemResponse = try await postJSON(
            url: redeemURL,
            body: RedeemRequest(token: token, solutions: solutions)
        )
        guard redeem.success, let capToken = redeem.token else {
            throw FaucetError.message("Captcha redeem rejected")
        }
        return capToken
    }

    private struct RedeemRequest: Encodable {
        let token: String
        let solutions: [Int]
    }

    // MARK: faucet POST

    private func postFaucet(
        address: String,
        capToken: String,
        amount: Double
    ) async -> TestnetFaucetOutcome {
        let url = host.appendingPathComponent("api/core-faucet")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body = ["address": address, "capToken": capToken]
        guard let payload = try? JSONSerialization.data(withJSONObject: body) else {
            return .failed(reason: "Failed to encode request")
        }
        request.httpBody = payload

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failed(reason: "Invalid response")
            }

            switch http.statusCode {
            case 200:
                if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let txid = json["txid"] as? String {
                    return .sent(txid: txid, amount: amount)
                }
                return .failed(reason: "Unexpected success payload")
            case 429:
                return .rateLimited(
                    message: "Rate limited (3/hour per IP). Try the web faucet."
                )
            default:
                let detail = parseErrorDetail(from: data)
                return .failed(
                    reason: "Faucet error \(http.statusCode): \(detail)"
                )
            }
        } catch {
            return .failed(reason: "Network error: \(error.localizedDescription)")
        }
    }

    /// Pulls the human-readable error out of a `{"detail":{"error":"..."}}`
    /// (or a flat `{"detail":"..."}`) body, falling back to a truncated raw
    /// body.
    private func parseErrorDetail(from data: Data) -> String {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let detail = json["detail"] as? [String: Any],
               let err = detail["error"] as? String {
                return err
            }
            if let detail = json["detail"] as? String {
                return detail
            }
        }
        return String(decoding: data.prefix(80), as: UTF8.self)
    }

    // MARK: JSON helpers

    private enum FaucetError: Error, LocalizedError {
        case message(String)
        var errorDescription: String? {
            switch self { case .message(let m): return m }
        }
    }

    private func postJSON<Body: Encodable, Out: Decodable>(
        url: URL,
        body: Body
    ) async throws -> Out {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode != 200 {
            throw FaucetError.message("HTTP \(http.statusCode)")
        }
        return try JSONDecoder().decode(Out.self, from: data)
    }
}

File: packages/kotlin-sdk/KotlinExampleApp/app/src/main/java/org/dashfoundation/example/services/faucet/TestnetFaucetService.kt
package org.dashfoundation.example.services.faucet

import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest

/**
 * **Dev/QA tool only.** App-level client for the testnet DASH faucet at
 * `faucet.thepasta.org` — port of `TestnetFaucetService.swift` (issue #3897).
 *
 * A plain `HttpURLConnection` client plus a pure-Kotlin `cap.js`
 * proof-of-work solver. It deliberately does **not** touch the wallet /
 * Platform FFI, `platform-wallet`, or any SDK business logic, so it lives in
 * the app layer, not the SDK. It lets QA top up a fresh/drained testnet
 * wallet with one tap instead of round-tripping through the web faucet.
 *
 * The faucet enforces a `cap.js` captcha; the soft challenge is a small
 * proof-of-work solved on-device off the main thread. The hard challenge
 * (returned on rate limit) is too expensive to brute-force here, so on rate
 * limit or any failure the caller falls back to the web faucet.
 */

/** cap.js proof-of-work solver (bit-identical to `@cap.js/widget`). */
internal object CapSolver {
    /**
     * cap.js seeded PRNG: FNV-1a-style seed + xorshift32. Emits [length]
     * lowercase hex chars. All arithmetic is unsigned 32-bit ([UInt] wraps).
     */
    fun prng(seed: String, length: Int): String {
        var state = 2166136261u // 0x811C9DC5
        for (ch in seed) { // seeds are ASCII (hex token + digits [+ "d"])
            state = state xor ch.code.toUInt()
            val sh = (state shl 1) + (state shl 4) + (state shl 7) + (state shl 8) + (state shl 24)
            state += sh
        }
        val out = StringBuilder()
        while (out.length < length) {
            state = state xor (state shl 13)
            state = state xor (state shr 17)
            state = state xor (state shl 5)
            // Zero-extended unsigned value → 8 lowercase hex chars.
            out.append(state.toLong().toString(16).padStart(8, '0'))
        }
        return out.substring(0, length)
    }

    /**
     * Brute-force the smallest non-negative `nonce` such that
     * lowercase-hex(`SHA256(salt + nonce)`) starts with [target]. Compares
     * against the raw digest bytes rather than materializing the hex string.
     * Suspend: checks cancellation every 0x10000 attempts so the caller's
     * `withTimeout` can actually preempt a slow challenge (← Swift
     * `CapSolver.solve`'s `Task.checkCancellation()` cadence — a worst-case
     * solve must not pin a core after the sheet already gave up).
     */
    suspend fun solve(salt: String, target: String): Long {
        val fullBytes = target.length / 2
        val hasNibble = target.length % 2 == 1
        val prefix = ByteArray(fullBytes) { i ->
            ((hexNibble(target[i * 2]) shl 4) or hexNibble(target[i * 2 + 1])).toByte()
        }
        val nibble = if (hasNibble) hexNibble(target[fullBytes * 2]) else 0

        val saltBytes = salt.toByteArray(Charsets.UTF_8)
        val md = MessageDigest.getInstance("SHA-256")
        var nonce = 0L
        while (true) {
            md.reset()
            md.update(saltBytes)
            md.update(nonce.toString().toByteArray(Charsets.UTF_8))
            val digest = md.digest()
            if (digestHasPrefix(digest, prefix, nibble, hasNibble)) return nonce
            nonce++
            if (nonce and 0xFFFF == 0L) {
                kotlin.coroutines.coroutineContext.ensureActive()
            }
        }
    }

    private fun digestHasPrefix(
        digest: ByteArray,
        prefix: ByteArray,
        nibble: Int,
        hasNibble: Boolean,
    ): Boolean {
        for (i in prefix.indices) {
            if (digest[i] != prefix[i]) return false
        }
        if (hasNibble) {
            return ((digest[prefix.size].toInt() and 0xFF) ushr 4) == nibble
        }
        return true
    }

    private fun hexNibble(c: Char): Int = when (c) {
        in '0'..'9' -> c - '0'
        in 'a'..'f' -> c - 'a' + 10
        in 'A'..'F' -> c - 'A' + 10
        else -> 0
    }
}

/** Result of a faucet funding attempt. */
sealed interface TestnetFaucetOutcome {
    /** Funds sent. [txid] is the Core tx id; [amount] the tDASH amount. */
    data class Sent(val txid: String, val amount: Double) : TestnetFaucetOutcome

    /** Server rate-limited (429). */
    data class RateLimited(val message: String) : TestnetFaucetOutcome

    /** Any other failure (network, solve, non-200). */
    data class Failed(val reason: String) : TestnetFaucetOutcome
}

/** Thin suspend client for the testnet faucet. Stateless; create one per use. */
class TestnetFaucet {
    private val host = "https://faucet.thepasta.org"
    private val json = Json { ignoreUnknownKeys = true }

    @Serializable
    private data class FaucetStatus(val capEndpoint: String, val coreFaucetAmount: Double)

    @Serializable
    private data class Challenge(val c: Int, val s: Int, val d: Int)

    @Serializable
    private data class ChallengeResponse(val challenge: Challenge, val token: String)

    @Serializable
    private data class RedeemRequest(val token: String, val solutions: List<Long>)

    @Serializable
    private data class RedeemResponse(val success: Boolean, val token: String? = null)

    @Serializable
    private data class FaucetResult(val txid: String? = null)

    /**
     * Request ~1 tDASH to [address]. Solves the soft captcha on-device and
     * posts to `/api/core-faucet`. Never throws — always maps to an outcome.
     */
    suspend fun requestCoreDash(address: String): TestnetFaucetOutcome {
        val trimmed = address.trim()
        if (trimmed.isEmpty()) return TestnetFaucetOutcome.Failed("No address available")

        val status = try {
            getJson("$host/api/status", FaucetStatus.serializer())
        } catch (e: Exception) {
            return TestnetFaucetOutcome.Failed("Faucet status unavailable: ${e.message}")
        }

        // Defense-in-depth: only solve against an HTTPS endpoint on the
        // faucet's own registrable domain, so a tampered response can't
        // downgrade the exchange or repoint the address-carrying POSTs.
        val capBase = try {
            URL(status.capEndpoint)
        } catch (e: Exception) {
            return TestnetFaucetOutcome.Failed("Untrusted captcha endpoint")
        }
        if (capBase.protocol != "https" ||
            !sameRegistrableDomain(capBase.host, URL(host).host)
        ) {
            return TestnetFaucetOutcome.Failed("Untrusted captcha endpoint")
        }

        val capToken = try {
            solveCaptcha(status.capEndpoint)
        } catch (e: Exception) {
            return TestnetFaucetOutcome.Failed("Captcha solve failed: ${e.message}")
        }

        return postFaucet(trimmed, capToken, status.coreFaucetAmount)
    }

    /** cap.js handshake: challenge → brute-force the sub-challenges → redeem. */
    private suspend fun solveCaptcha(capEndpoint: String): String {
        val base = capEndpoint.trimEnd('/')
        val challenge = postJson(
            "$base/challenge",
            "{}",
            ChallengeResponse.serializer(),
        )
        val c = challenge.challenge.c
        val s = challenge.challenge.s
        val d = challenge.challenge.d
        val token = challenge.token

        // Reject pathological challenge sizes up front (a bad/future faucet
        // response could otherwise pin every core with no in-app abort).
        require(c in 1..256 && s in 1..256 && d in 1..6) {
            "Unsupported captcha challenge (c=$c, s=$s, d=$d)"
        }

        val solutions: List<Long> = withTimeout(30_000) {
            coroutineScope {
                (1..c).map { i ->
                    async(Dispatchers.Default) {
                        val salt = CapSolver.prng("$token$i", s)
                        val target = CapSolver.prng("$token${i}d", d)
                        CapSolver.solve(salt, target)
                    }
                }.awaitAll()
            }
        }

        val redeem = postJson(
            "$base/redeem",
            json.encodeToString(RedeemRequest.serializer(), RedeemRequest(token, solutions)),
            RedeemResponse.serializer(),
        )
        require(redeem.success && redeem.token != null) { "Captcha redeem rejected" }
        return redeem.token
    }

    /** POSTs the funding request and maps the response to an outcome. */
    private suspend fun postFaucet(
        address: String,
        capToken: String,
        amount: Double,
    ): TestnetFaucetOutcome = withContext(Dispatchers.IO) {
        try {
            val conn = openPost("$host/api/core-faucet")
            val body = json.encodeToString(
                CoreFaucetRequest.serializer(),
                CoreFaucetRequest(address, capToken),
            )
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            when (val code = conn.responseCode) {
                200 -> {
                    val txid = json.decodeFromString(
                        FaucetResult.serializer(),
                        conn.inputStream.bufferedReader().use { it.readText() },
                    ).txid
                    if (txid != null) {
                        TestnetFaucetOutcome.Sent(txid, amount)
                    } else {
                        TestnetFaucetOutcome.Failed("Unexpected success payload")
                    }
                }
                429 -> TestnetFaucetOutcome.RateLimited(
                    "Rate limited (3/hour per IP). Try the web faucet.",
                )
                else -> {
                    val detail = conn.errorStream?.bufferedReader()?.use { it.readText() }
                        ?.take(120) ?: ""
                    TestnetFaucetOutcome.Failed("Faucet error $code: $detail")
                }
            }
        } catch (e: Exception) {
            TestnetFaucetOutcome.Failed("Network error: ${e.message}")
        }
    }

    @Serializable
    private data class CoreFaucetRequest(val address: String, val capToken: String)

    // --- HTTP helpers -------------------------------------------------------

    private suspend fun <T> getJson(
        url: String,
        deserializer: kotlinx.serialization.DeserializationStrategy<T>,
    ): T = withContext(Dispatchers.IO) {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 30_000
            readTimeout = 30_000
        }
        if (conn.responseCode != 200) throw Exception("HTTP ${conn.responseCode}")
        json.decodeFromString(deserializer, conn.inputStream.bufferedReader().use { it.readText() })
    }

    private suspend fun <T> postJson(
        url: String,
        body: String,
        deserializer: kotlinx.serialization.DeserializationStrategy<T>,
    ): T = withContext(Dispatchers.IO) {
        val conn = openPost(url)
        conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
        if (conn.responseCode != 200) throw Exception("HTTP ${conn.responseCode}")
        json.decodeFromString(deserializer, conn.inputStream.bufferedReader().use { it.readText() })
    }

    private fun openPost(url: String): HttpURLConnection =
        (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 30_000
            readTimeout = 30_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }

    companion object {
        /** Web-faucet URL, used as the fallback when the API path fails. */
        const val WEB_URL = "https://faucet.thepasta.org/"

        /**
         * True when two hosts share their registrable domain (last two DNS
         * labels), e.g. `cap.thepasta.org` and `faucet.thepasta.org`.
         */
        private fun sameRegistrableDomain(a: String, b: String): Boolean {
            fun base(h: String) = h.split(".").takeLast(2).joinToString(".")
            val ba = base(a)
            return ba.contains(".") && ba == base(b)
        }
    }
}
